packages/@ember/-internals/extension-support/lib/data_adapter.js

```diff
@@ -284,28 +409,33 @@ export default EmberObject.extend({
     let klass = this._nameToClass(modelName);
     let records = this.getRecords(klass, modelName);
 
-    function onChange() {
+    let onChange = () => {
       typesUpdated([this.wrapModelType(klass, modelName)]);
-    }
+    };
 
-    let observer = {
-      didChange(array, idx, removedCount, addedCount) {
-        // Only re-fetch records if the record count changed
-        // (which is all we care about as far as model types are concerned).
-        if (removedCount > 0 || addedCount > 0) {
-          scheduleOnce('actions', this, onChange);
+    let { typeWatchers } = this;
+
+    let typeWatcher = typeWatchers.get(records);
+
+    if (!typeWatcher) {
+      if (typeWatchers.size === 0) {
+        backburner.on('end', this.flushWatchers);
```

Comment:
We should only add `on('end', this.flushWatchers)` once (seems like we might add it twice).

Reply:
ah, you're correct. I'll extract this out to a shared function